nodego/supervisor.go

```diff
@@ -0,0 +1,255 @@
+		return nil
+	}
+
+	if err := postToSupervisor("/_ah/log", batch, supervisorLogTimeout); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func flushLogBatches() {
+	select {
+	case logBatch := <-logBatchesToReport:
+		go flushLogBatches()
+		if err := logBatch.report(); err != nil {
+			fmt.Fprintln(os.Stderr, err.Error())
+			killInstance()
+		}
+	default:
+		startNewLogBatch()
+	}
+}
+
+func logBatchReportWorker() {
+	for logBatch := range logBatchesToReport {
+		// TODO it may be nice to have a time base buffering method on top of this.
+		<-logBatch.ready
+
+		loggingContext.Lock()
+		if logBatch == loggingContext.currentLogBatch {
+			startNewLogBatch()
+		}
+		loggingContext.Unlock()
+
+		if err := logBatch.report(); err != nil {
+			fmt.Fprintln(os.Stderr, err.Error())
+			killInstance()
+		}
+	}
+}
+
+func newSupervisorRequest(path string, v interface{}) (*http.Request, error) {
+	postData, err := json.Marshal(v)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequest("POST", (&url.URL{
+		Scheme: "http",
+		Host:   supervisorHostname + ":" + supervisorInternalPort,
+		Path:   path,
+	}).String(), bytes.NewBuffer(postData))
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Add("Content-Type", "application/json")
+	req.Header.Add("Content-Length", strconv.Itoa(len(postData)))
+
+	return req, nil
+}
+
+func postToSupervisor(path string, v interface{}, timeout time.Duration) error {
+	ctx, cancel := context.WithTimeout(context.Background(), timeout)
+	defer cancel()
+
+	req, err := newSupervisorRequest(path, v)
+	if err != nil {
+		return err
+	}
+
+	resp, err := ctxhttp.Do(ctx, nil, req)
+	if err == ctx.Err() {
+		return errors.New("timeout when calling supervisor")
+	} else if err != nil {
+		return fmt.Errorf("error when calling supervisor: %s\n\n%s\n", err.Error(), debug.Stack())
+	}
+
+	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
+		err = fmt.Errorf("incorrect response code from supervisor: %d\n", resp.StatusCode)
+	}
+
+	return err
+}
+
+func killInstance() {
+	postToSupervisor("/_ah/kill", nil, supervisorKillTimeout)
+	os.Exit(16)
+}
+
+func loggerMiddleware(handler http.Handler) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		loggingContext.Lock()
+		loggingContext.functionExecutionID = r.Header.Get("Function-Execution-Id")
+		loggingContext.Unlock()
+
+		defer func() {
+			go flushLogBatches()
+			loggingContext.Lock()
+			loggingContext.functionExecutionID = ""
+			loggingContext.Unlock()
+		}()
+
+		handler.ServeHTTP(w, r)
+	}
+}
+
+// WithLogger returns an http handler that reads the function execution id,
+// attaches it to log messages sent to the supervisor and flushes pending logs
+// afterwards.
+func WithLogger(handler http.Handler) http.Handler {
+	return loggerMiddleware(handler)
+}
+
+// WithLoggerFunc is the same as WithLogger but accepts a handler function.
+func WithLoggerFunc(handler func(w http.ResponseWriter, r *http.Request)) http.HandlerFunc {
```

Comment:
You should be able to just accept a http.HandlerFunc and then not need to cast it below.